arch/risc-v/src/mpfs/mpfs_ethernet.c

```diff
@@ -0,0 +1,3841 @@
+    }
+
+  /* Wait for the PHY reset to complete */
+
+  ret = -ETIMEDOUT;
+  for (timeout = 0; timeout < PHY_RESET_WAIT_COUNT; timeout++)
+    {
+      mcr = GMII_MCR_RESET;
+      int result = mpfs_phyread(priv, priv->phyaddr,
+                                GMII_MCR, &mcr);
+      if (result < 0)
+        {
+          nerr("ERROR: Failed to read the MCR register: %d\n", ret);
+          ret = result;
+        }
+      else if ((mcr & GMII_MCR_RESET) == 0)
+        {
+          ret = OK;
+          break;
+        }
+    }
+
+  /* Disable management port */
+
+  mpfs_disablemdio(priv);
+  return ret;
+}
+
+/****************************************************************************
+ * Function: mpfs_ethconfig
+ *
+ * Description:
+ *  Configure the Ethernet interface for DMA operation.
+ *
+ * Parameters:
+ *   priv - A reference to the private driver state structure
+ *
+ * Returned Value:
+ *   OK on success; Negated errno on failure.
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+static int mpfs_ethconfig(struct mpfs_ethmac_s *priv)
+{
+  int ret;
+
+  ninfo("Entry\n");
+
+#ifdef CONFIG_MPFS_PHYINIT
+  /* Perform any necessary, board-specific PHY initialization */
+
+  ret = mpfs_phy_boardinitialize(0);
+  if (ret < 0)
+    {
+      nerr("ERROR: Failed to initialize the PHY: %d\n", ret);
+      return ret;
+    }
+#endif
+
+  /* Reset the Ethernet block */
+
+  ninfo("Reset the Ethernet block\n");
+  mpfs_ethreset(priv);
+
+  /* Initialize the PHY */
+
+  ninfo("Initialize the PHY\n");
+  ret = mpfs_phyinit(priv);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  /* Initialize the MAC and DMA */
+
+  ninfo("Initialize the MAC and DMA\n");
+  ret = mpfs_macconfig(priv);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  /* Enable normal MAC operation */
+
+  ninfo("Enable normal operation\n");
+  return mpfs_macenable(priv);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Function: mpfs_ethinitialize
+ *
+ * Description:
+ *   Initialize the Ethernet driver for one interface.  If the STM32 chip
+ *   supports multiple Ethernet controllers, then board specific logic
+ *   must implement arm_netinitialize() and call this function to initialize
+ *   the desired interfaces.
+ *
+ * Parameters:
+ *   intf - In the case where there are multiple EMACs, this value
+ *          identifies which GMAC is to be initialized.
+ *
+ * Returned Value:
+ *   OK on success; Negated errno on failure.
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+#if defined(CONFIG_NETDEV_LATEINIT)
+int mpfs_ethinitialize(int intf)
+#else
+static inline int mpfs_ethinitialize(int intf)
+#endif
```

Comment:
Please explain why we need this? I do not see that `mpfs_ethinitialize` is called out of this file. I mean either it should be `static` always or we should add `mpfs_ethinitialize(0);` calls to `mpfs_bringup.c` and function prototype should be exposed in the header, isn't it?

Reply:
I guess that the idea is that by defining CONFIG_NETDEV_LATEINIT you can call mpfs_ethinitialize in you own board files, and the call to the riscv_netinitialize is suppressed in the common code. This is the same as what is done for the arm boards - typically you just initialize it via the common riscv_initialize, but can make an exception by defining CONFIG_NETDEV_LATEINIT and doing your own initialization.

Of course, we could have the header file for these function prototypes (mpfs_ethinitialize and mpfs_phy_boardinitialize) similarly as what is done for arm boards (e.g. stm32_ethernet.h).

Reply:
Lets got with `int mpfs_ethinitialize(int intf);` only and remove `static` case